src/pre_process_data.py

```diff
@@ -1,71 +1,95 @@
-# author: Group 24
-# date: 2021-11-25
-
-'''This script creates pre-processor for the Crime Vancouver data (from https://geodash.vpd.ca/opendata/)
-
-Usage: pre_process_data.py --out_path=<out_path>
-
-Options:
---out_path=<out_path>       Path to directory where the processed data and pre-processor object should be written
-'''
-import pandas as pd
-import numpy as np
-from docopt import docopt
-import pickle
-from sklearn.model_selection import train_test_split
-from sklearn.compose import ColumnTransformer, make_column_transformer
-from sklearn.pipeline import Pipeline, make_pipeline
-from sklearn.impute import SimpleImputer
-from sklearn.preprocessing import (
-    OneHotEncoder,
-    OrdinalEncoder,
-    PolynomialFeatures,
-    StandardScaler,
-)
-
-opt = docopt(__doc__)
-
-def main(out_path):
-   
-    # Define columns for transformation
-    drop_features = ["HUNDRED_BLOCK"]
-    neighbour_feature = ["NEIGHBOURHOOD"]
-    categorical_features = ["YEAR", "MONTH", "DAY", "HOUR", "MINUTE"]
-    numeric_features = ["X", "Y"]
-
-    # preprocessor for EDA and model training
-    preprocessor = make_column_transformer(
-        (
-            make_pipeline(
-                SimpleImputer(strategy="constant", fill_value="Central Business District"),
-                OneHotEncoder(handle_unknown="ignore", sparse=False),
-            ),
-            neighbour_feature,
-        ),
-        (OneHotEncoder(handle_unknown="ignore", sparse=False), categorical_features),
-       (
-           make_pipeline(
-               SimpleImputer(strategy="median"),
-               StandardScaler(),
-           ),
-           numeric_features
-       ),
-        ("drop", drop_features),
-    )
-    
-
-    # Save pre-processed data        
-    try:
-        filename = 'preprocessor.p'
-        outfile = open(out_path+ "/" + filename,'wb')
-        pickle.dump(preprocessor, outfile)
-        outfile.close()
-        
-        print(f"Pre-processor is loaded successfully at %s" %(out_path + filename))    
-        
-    except Exception as error:
-        print(f"Error message: %s" %error)
-        print("Error while saving pre-processor!")
-    
-if __name__ == "__main__":
-    main(opt['--out_path'])
+# author: Group 24

```

Comment:
Hi Ramiro,

Since it is a jupyter notebook I put my comments here instead =)

1. best_LR_model. In case you want to automate it to get the best model automatically, you should then return your results from cv_models as a dict of {"cv_result": results_df, "best_model" model_object}. You can get the highest score from the table and then you can filter out the best model from "models" object.

However that creates another problem, you still have to have specific input for hyperparameterization - technically can be automated with a predefined dict of possible options, HOWEVER I would not recommend to go for that path ... not worth it for now.

So in summary, I think at this moment it is good!
2. I will merge my balance of HOUR back to this script instead.

3. We need to extract more graphs and tables from these steps so that we can answer the following in the report

- What pre-processing actions we have made and why? 
  - A graph that shows the distribution of hours ( HOUR == 0 is crazily high) <=== I will do it
  - A graph that shows distribution after HOUR is balanced <==== I will do it
- A table of the CVs among different models (which we have, just need to be in a png)
- The results (confusion matrix, .... etc)

Reply:
Answer 1.:

I agree with that, and I was thinking the same but I still don't understand how will use these outputs in the document. We can get outputs from the functions and then used them as arguments for the next functions, but this depends on the report document.

For the hyperparameter optimization I hard coded the parameter grid I did not want to do that but at this point, we don't have many choices 😭, the time complexity is a problem.

Answer 3.

I agree let's work on this together, we have to use the outputs to describe our workflow.

Reply:
As we talked, hardcode it for now for this project =D, and we can output the pipe_best for people who wants to run the prediction themselves lol